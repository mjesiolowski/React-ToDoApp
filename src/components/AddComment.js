import React, { useState } from 'react'
import { connect } from 'react-redux'
import { addComment } from '../actions/tasks'
import { commentAlert as commentAlertAction } from '../actions/alerts'

export const AddComment = ({ addComment, commentAlertAction, taskId, commentAlert }) => {

   const [commentValue, setCommentValue] = useState("")

   const handleSubmit = (e) => {
      e.preventDefault()
      if (commentValue.length > 0) {
         addComment(taskId, commentValue)
         commentAlertAction(false)
         setCommentValue("")
      }
      else {
         commentAlertAction(true)
      }
   }

   return (
      <form
         onSubmit={handleSubmit}
         className="form"
      >
         <input
            value={commentValue}
            onChange={(e) => setCommentValue(e.target.value)}
            type="text"
            placeholder="Add comment"
            className="form__input" />

         <button className="task__button button">Add comment</button>

         {commentAlert && <p className='text'>Minimum length of 1 character required</p>}
      </form>

   )
}

const mapDispatchToProps = (dispatch) => ({
   addComment: (taskId, commentValue) => dispatch(addComment(taskId, commentValue)),
   commentAlertAction: (isTrue) => dispatch(commentAlertAction(isTrue))
}
)
const mapStateToProps = ({ alerts }) => {
   return {
      commentAlert: alerts.commentAlert
   }
}

export default connect(mapStateToProps, mapDispatchToProps)(AddComment)