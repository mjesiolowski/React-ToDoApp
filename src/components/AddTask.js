import React, { useState } from 'react';
import { connect } from 'react-redux'
import { addTask } from '../actions/tasks'
import moment from 'moment'
import { validateTask } from '../functions/validateTask'
import { handleAlerts } from '../functions/handleAlerts'

const AddTask = ({ dispatch, alerts, tasks }) => {
   const dateFormat = "DD.MM.YYYY"

   const [taskName, setTaskName] = useState("")
   const [taskPriority, setTaskPriority] = useState(false)
   const [taskDeadline, setTaskDeadline] = useState(moment().format(dateFormat))

   const task = {
      taskName,
      taskDeadline,
      dateFormat
   }

   const handleAddTask = (e, dateFormat) => {
      e.preventDefault()
      const addTaskAction = () => dispatch(addTask({
         name: taskName,
         isPriority: taskPriority,
         deadline: moment(taskDeadline, dateFormat).valueOf()
      }))

      const isTaskValid = handleAlerts(validateTask({ ...task }, tasks), dispatch)
      if (isTaskValid) {
         addTaskAction()
         setTaskName("")
      }
   }

   return (
      <section className="addTask">
         <h2 className="title__sub addTask__title">AddTask</h2>
         <form
            onSubmit={(e) => handleAddTask(e, dateFormat)}
            className='form'
         >
            <input
               type="text"
               placeholder="Add task"
               value={taskName}
               onChange={(e) => setTaskName(e.target.value)}
               className='form__input'
            />

            <input
               value={taskDeadline}
               onChange={(e) => setTaskDeadline(e.target.value)}
               className='form__input'
            />

            <div className="form__wrapper">
               <label
                  htmlFor="priority"
                  className='form__label'
               >Priority: </label>

               <select
                  id="priority"
                  value={taskPriority}
                  onChange={(e) => setTaskPriority(e.target.value === "true" ? true : false)}
                  className='form__select'
               >
                  <option value="false">Not important</option>
                  <option value="true">Important</option>

               </select>
            </div>

            {alerts.duplicateAlert && <p className='addTask__alert text'>Task already on the to do list</p>}
            {alerts.lengthAlert && <p className='addTask__alert text'>Minimum 3 characters required</p>}
            {alerts.dateAlert && <p className='addTask__alert text'>Your deadline date must not be set in the past. Date fromat: DD.MM.YYYY
               </p>}
            <button className='addTask__button button'><i className="fas fa-plus-circle"></i></button>
         </form>
      </section>
   )
}

const mapStateToProps = ({ tasks, alerts }) => ({
   tasks,
   alerts
})

export default connect(mapStateToProps)(AddTask)