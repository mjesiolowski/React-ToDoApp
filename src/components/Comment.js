import React, { useState } from 'react';
import { connect } from 'react-redux'
import moment from 'moment'
import { editComment, deleteComment } from '../actions/tasks'

export const Comment = ({ editComment, deleteComment, comment, id: taskId, completed }) => {

   const [commentText, setCommentText] = useState(comment.text)

   const handleEditComment = () => {
      editComment(taskId, { ...comment, beingEdited: true, })
   }


   const handleDeleteComment = () => {
      deleteComment(taskId, comment.id)
   }

   const handleSubmit = (e) => {
      e.preventDefault()
      commentText.length > 0 ?
         editComment(taskId,
            {
               ...comment,
               beingEdited: false,
               text: commentText
            })
         :
         editComment(taskId, { ...comment, beingEdited: false, })
   }

   return (
      <>
         {comment.beingEdited ?
            <form
               onSubmit={handleSubmit}
               className="form"
            >
               <input
                  type="text"
                  value={commentText}
                  onChange={(e) => setCommentText(e.target.value)}
                  className="form__input"
               />
               <p className="text">{moment(comment.createdAt).format('DD-MM-YYYY')}</ p>
            </form>

            :

            <div className="task__comment">
               <p className="task__text text">{commentText}</p>
               <p className="task__text text">{moment(comment.createdAt).format('DD-MM-YYYY')}</ p>

               {!completed && <button
                  onClick={handleEditComment}
                  className="task__button button">Edit comment</button>}

               {!completed && <button
                  onClick={handleDeleteComment}
                  className="task__button button">Delete comment</button>}
            </div>
         }


      </>

   )
}

const mapDispatchToProps = (dispatch) => ({
   editComment: (id, update) => dispatch(editComment(id, update)),
   deleteComment: (id, update) => dispatch(deleteComment(id, update))
})

export default connect(undefined, mapDispatchToProps)(Comment)