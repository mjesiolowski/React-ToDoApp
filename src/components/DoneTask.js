import React from 'react';
import { connect } from 'react-redux'
import moment from 'moment'
import RenderComments from './RenderComments'


export const DoneTask = ({ task }) => {
   const dateFormat = "DD.MM.YYYY"


   return (
      <section className="task">
         <div className="container">
            <h1 className="title__sub">{task.name}</h1>
            <p className="task__text text">Created at: {moment(task.createdAt).format(dateFormat)}</p>
            <p className="task__text text">Completed at: {moment(task.completedAt).format(dateFormat)}</p>
         </div>

         <div className="comments">
            <ul className="list">
               <h2 className="title__sub">
                  Comments:
               </h2>
               <RenderComments task={task} />
            </ul>
         </div>

      </section>
   )
}

const mapStateToProps = (state, props) => {
   return {
      task: state.tasks.find(({ id }) => id === props.match.params.id)
   }
}


export default connect(mapStateToProps)(DoneTask)