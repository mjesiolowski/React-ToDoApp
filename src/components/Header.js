import React from 'react';
import { NavLink } from 'react-router-dom'
import HomePage from './HomePage'

const Header = () => {

   return (
      <>
         <NavLink to="/"><h1>ToDoList App</h1></NavLink>
      </>
   )
}

export default Header