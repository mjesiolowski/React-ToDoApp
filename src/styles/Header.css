.header {
   margin: 0 auto;
}

.header__title {
   border-bottom: 1px solid #61DAF7;
   font-size: 4rem;
   letter-spacing: .2rem;
   line-height: 15rem;
}

.header__time {
   margin: .7rem;
   font-size: 20px;
}

.header__form {
   display: flex;
   flex-wrap: wrap;
   justify-content: space-between;
   align-items: flex-start;
   width: 70%;
   padding: 0 2rem 0 2rem;
   margin: 2rem auto 0;
}

.header__input {
   flex-grow: 1;
   height: 4rem;
   padding: 0 1.5rem;
   margin-bottom: 2.5rem;
   border: none;
   background: #DDD;
   font-size: 1.6rem;
}

.header__icon {
   margin: 0 .7rem;
   font-size: 3rem;
   line-height: 4rem;
   cursor: pointer;
   transition: .2s;
}

.header__icon:hover {
   color: #61DAF7;
   transform: scale(1.2)
}

.header__warning-info {
   margin: 2rem auto;
   font-size: 1.8rem;
}

@media (max-width: 1050px) {
   .header__form {
      justify-content: space-evenly;
   }

   .header__input {
      flex-basis: 100%
   }

   .header__icon {
      margin: .7rem;
   }
}

@media (max-width: 750px) {
   .header__title {
      font-size: 3.5rem;
      letter-spacing: .3rem;
      line-height: 10rem;
   }
}