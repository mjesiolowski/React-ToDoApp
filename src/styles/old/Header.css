.header__title {
   border-bottom: 1px solid #61DAF7;
   font-size: 4rem;
   letter-spacing: .2rem;
   line-height: 15rem;
}

.header__time {
   margin: .7rem;
   font-size: 20px;
}

.header__form {
   display: flex;
   flex-wrap: wrap;
   justify-content: space-evenly;
   align-items: center;
   width: 70%;
   padding: 0 2rem 0 2rem;
   margin: 2rem auto 0;
}

.header__form__input {
   flex-basis: 100%;
   height: 4rem;
   padding: 0 1.5rem;
   margin-bottom: 1rem;
   border: none;
   background: #DDD;
   font-size: 1.6rem;
}

.header__form__icon {
   margin: 1rem .7rem;
   font-size: 3rem;
   line-height: 4rem;
   cursor: pointer;
   transition: .2s;
}

.header__icon:hover {
   color: #61DAF7;
   transform: scale(1.2)
}

.header__warning-info {
   margin: 2rem auto;
   font-size: 1.8rem;
}


@media (max-width: 750px) {
   .header__title {
      font-size: 3.5rem;
      letter-spacing: .3rem;
      line-height: 10rem;
   }
}