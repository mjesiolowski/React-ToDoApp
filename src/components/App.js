import React, { Component } from 'react';
import Header from './Header'
import TasksList from './TasksList'
import uuid from 'uuid'


class App extends Component {
  state = {
    editedInputValue: '',
    headerInputValue: '',
    minimalTaskLengthAlert: false,
    searchSectionActive: false,
    searchedTasks: [],
    taskDuplicatedAlert: false,
    tasks: [
      {
        id: uuid(),
        text: "Task 1",
        active: true,
        edited: false,
      },
      {
        id: uuid(),
        text: "Task 2",
        active: true,
        edited: false,
      },
      {
        id: uuid(),
        text: "Task 3",
        active: true,
        edited: false,
      },
    ],
  }

  //INPUT HANDLERS

  handleHeaderInputValue = e => {
    this.setState({
      headerInputValue: e.target.value
    })
  }

  handleHeaderInputSubmit = e => {
    e.preventDefault()
    this.handleAddTaskButton()
  }

  handleEditInputValue = e => {
    this.setState(({
      editedInputValue: e.target.value
    }))
  }

  //ADDING TASK
  // handleAddTaskButton = () => {
  //   const tasksToDo = [...this.state.tasks]
  //   const duplicateCheck = tasksToDo.filter(task => task.text === this.state.headerInputValue && task.active).length

  //   if (this.state.headerInputValue.length <= 2) {
  //     this.setState({
  //       minimalTaskLengthAlert: true
  //     })
  //   }

  //   else if (duplicateCheck) {
  //     this.setState({
  //       taskDuplicatedAlert: true
  //     })
  //   }
  //   else {
  //     tasksToDo.push(
  //       {
  //         id: uuid(),
  //         text: this.state.headerInputValue,
  //         active: true,
  //         edited: false,
  //       }
  //     )
  //     this.setState({
  //       tasks: tasksToDo,
  //       headerInputValue: ''
  //     })
  //   }
  // }

  //CHANGING TASK STATUS
  handleDoneTaskButton = id => {
    const tasks = [...this.state.tasks]
    tasks.forEach(task => {
      if (task.id === id) {
        task.active = !task.active
        task.time = new Date().getTime()
      }
    })
    this.setState({
      tasks,
    })
  }

  //EDITING TASK
  handleEditButton = (id, task) => {
    if (!this.state.editedInputValue) {
      this.setState({
        editedInputValue: task
      })

      const tasks = [...this.state.tasks]
      tasks.forEach(task => {
        if (task.id === id) {
          task.edited = !task.edited
        }
      })
      this.setState({
        tasks,
      })
    }
  }

  handleEditingTask = (e, id) => {
    e.preventDefault()
    const tasks = [...this.state.tasks]

    if (this.state.editedInputValue.length < 3) { alert('Task too short') }
    else {
      tasks.map(task => {
        if (task.id === id) {
          task.text = this.state.editedInputValue
          task.edited = false
        } return null
      })

      this.setState(() => ({
        editedInputValue: '',
        tasks,
      }))
    }
  }

  //REMOVING TASK
  handleRemoveTaskButton = id => {
    const tasks = [...this.state.tasks]
    const tasksAfterRemoval = tasks.filter((task) => {
      if (task.id !== id) {
        return task
      } return null
    })

    this.setState({
      tasks: tasksAfterRemoval,
    })
  }

  //SEARCHING TASK
  handleSearchTaskButton = () => {
    const tasks = [...this.state.tasks]

    if (this.state.headerInputValue.length) {
      const searchResult = tasks.filter((result) => result.text.toLowerCase().includes(this.state.headerInputValue.trim().toLowerCase()) && result.active
      )

      this.setState({
        headerInputValue: '',
        searchSectionActive: true,
        searchedTasks: [...searchResult]
      })
    }
  }

  handleReturnButton = () => {
    this.setState({
      searchSectionActive: false,
    })
  }

  //ALERT HANDLERS

  componentDidUpdate(prevProps, prevState) {
    if (this.state.minimalTaskLengthAlert && this.state.headerInputValue.length > 2) {
      this.setState({
        minimalTaskLengthAlert: false
      })
    }
    else if (prevState.headerInputValue !== this.state.headerInputValue) {
      this.setState({
        taskDuplicatedAlert: false
      })
    }
  }

  //RENDERING

  render() {
    return (
      <div className="App">

        <header className="header">
          <Header
            alert={this.state.minimalTaskLengthAlert}
            duplicate={this.state.taskDuplicatedAlert}
            value={this.state.headerInputValue}

            handleAddTaskButton={this.handleAddTaskButton}
            handleHeaderInputSubmit={this.handleHeaderInputSubmit}
            handleHeaderInputValue={this.handleHeaderInputValue}
            handleSearchTaskButton={this.handleSearchTaskButton}
          />
        </header>

        <main className="tasksList">
          <TasksList
            editedValue={this.state.editedInputValue}
            searchSection={this.state.searchSectionActive}
            searchedTasks={this.state.searchedTasks}
            tasksToDo={this.state.tasks}

            handleEditButton={this.handleEditButton}
            handleEditInputValue={this.handleEditInputValue}
            handleEditingTask={this.handleEditingTask}
            handleDoneTaskButton={this.handleDoneTaskButton}
            handleReturnButton={this.handleReturnButton}
            handleRemoveTaskButton={this.handleRemoveTaskButton}
          />
        </main>

      </div>
    );
  }
}

export default App;
