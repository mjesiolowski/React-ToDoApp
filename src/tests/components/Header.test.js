import React from 'react'
import { shallow } from 'enzyme'
import { Header } from '../../components/Header'

let wrapper, dispatch

beforeEach(() => {
  dispatch = jest.fn()

  wrapper = shallow(<Header
    dispatch={dispatch}
  />)
})

test('should render', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should handle return button', () => {
  wrapper.find('h1').simulate('click')
  expect(dispatch).toBeCalled()
})