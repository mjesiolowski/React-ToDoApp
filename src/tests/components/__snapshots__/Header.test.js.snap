// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<header
  className="header"
>
  <h1
    className="title"
    onClick={[Function]}
  >
       
    <NavLink
      className="link"
      to="/"
    >
      Task App
    </NavLink>
  </h1>
</header>
`;
